File: algo/copypasta/python/binary_search.py
"""在已排序序列中执行二分查找。"""

from collections.abc import Sequence
from typing import Protocol


class Comparable(Protocol):
    def __lt__(self, other: object, /) -> bool: ...
    def __le__(self, other: object, /) -> bool: ...


def lower_bound[T: Comparable](values: Sequence[T], target: T) -> int:
    """返回第一个值大于或等于 ``target`` 的下标；若不存在则返回序列长度。"""
    left, right = 0, len(values)
    while left < right:
        middle = (left + right) // 2
        if values[middle] < target:
            left = middle + 1
        else:
            right = middle
    return left


def upper_bound[T: Comparable](values: Sequence[T], target: T) -> int:
    """返回第一个值大于 ``target`` 的下标；若不存在则返回序列长度。"""
    left, right = 0, len(values)
    while left < right:
        middle = (left + right) // 2
        if values[middle] <= target:
            left = middle + 1
        else:
            right = middle
    return left


def binary_search[T: Comparable](values: Sequence[T], target: T) -> int | None:
    """返回 ``target`` 首次出现的下标；若不存在则返回 ``None``。"""
    index = lower_bound(values, target)
    return index if index < len(values) and values[index] == target else None

File: algo/copypasta/cpp/binary_search.cpp
#include <functional>

// 在半开区间 [left, right) 中寻找第一个使 predicate 为真的整数。
// predicate 必须单调地由 false 变为 true；若区间内都为 false，则返回 right。
int first_true(int left, int right, const std::function<bool(int)>& predicate) {
    while (left < right) {
        int middle = left + (right - left) / 2;
        if (predicate(middle)) {
            right = middle;
        } else {
            left = middle + 1;
        }
    }
    return left;
}

// 在闭区间 [left, right] 中寻找最后一个使 predicate 为真的整数。
// predicate 必须单调地由 true 变为 false，且调用前须保证 predicate(left) 为 true。
int last_true(int left, int right, const std::function<bool(int)>& predicate) {
    while (left < right) {
        int middle = left + (right - left + 1) / 2;
        if (predicate(middle)) {
            left = middle;
        } else {
            right = middle - 1;
        }
    }
    return left;
}
